src/chunk_constraint.c

```diff
@@ -1,34 +1,306 @@
 #include <utils/hsearch.h>
+#include <utils/relcache.h>
 #include <utils/rel.h>
+#include <utils/builtins.h>
+#include <utils/lsyscache.h>
+#include <utils/syscache.h>
 #include <access/heapam.h>
 #include <access/xact.h>
 #include <catalog/indexing.h>
 #include <catalog/pg_constraint.h>
+#include <catalog/pg_constraint_fn.h>
 #include <catalog/objectaddress.h>
+#include <catalog/dependency.h>
+#include <funcapi.h>
 
 #include "scanner.h"
 #include "chunk_constraint.h"
 #include "chunk_index.h"
+#include "dimension_vector.h"
 #include "dimension_slice.h"
 #include "hypercube.h"
 #include "chunk.h"
 #include "hypertable.h"
 #include "errors.h"
 #include "process_utility.h"
 
-#define create_chunk_constraint(chunk, constraint_oid)		\
-	DatumGetObjectId(CatalogInternalCall2(DDL_CREATE_CHUNK_CONSTRAINT, \
-						 Int32GetDatum(chunk->fd.id),	\
-						 ObjectIdGetDatum(constraint_oid)))
+#define DEFAULT_EXTRA_CONSTRAINTS_SIZE 4
 
-static inline ChunkConstraint *
-chunk_constraint_fill(ChunkConstraint *cc, HeapTuple tuple)
+#define CHUNK_CONSTRAINTS_SIZE(num_constraints)		\
+	((sizeof(ChunkConstraint) * num_constraints))
+
+ChunkConstraints *
+chunk_constraints_alloc(int size_hint)
+{
+	ChunkConstraints *ccs = palloc(sizeof(ChunkConstraints));
+
+	ccs->capacity = size_hint + DEFAULT_EXTRA_CONSTRAINTS_SIZE;
+	ccs->num_constraints = 0;
+	ccs->num_dimension_constraints = 0;
+	ccs->constraints = palloc0(CHUNK_CONSTRAINTS_SIZE(ccs->capacity));
+
+	return ccs;
+}
+
+ChunkConstraints *
+chunk_constraints_copy(ChunkConstraints *ccs)
+{
+	ChunkConstraints *copy = palloc(sizeof(ChunkConstraints));
+
+	memcpy(copy, ccs, sizeof(ChunkConstraints));
+	copy->constraints = palloc0(CHUNK_CONSTRAINTS_SIZE(ccs->capacity));
+	memcpy(copy->constraints, ccs->constraints, CHUNK_CONSTRAINTS_SIZE(ccs->num_constraints));
+
+	return copy;
+}
+
+static void
+chunk_constraints_expand(ChunkConstraints *ccs, int16 new_capacity)
 {
-	memcpy(&cc->fd, GETSTRUCT(tuple), sizeof(FormData_chunk_constraint));
+	if (new_capacity <= ccs->capacity)
+		return;
+
+	ccs->capacity = new_capacity;
+	ccs->constraints = repalloc(ccs->constraints, CHUNK_CONSTRAINTS_SIZE(new_capacity));
+}
+
+static ChunkConstraint *
+chunk_constraints_add(ChunkConstraints *ccs,
+					  int32 chunk_id,
+					  int32 dimension_slice_id,
+					  const char *constraint_name,
+					  const char *hypertable_constraint_name)
+{
+	ChunkConstraint *cc;
+
+	chunk_constraints_expand(ccs, ccs->num_constraints + 1);
+	cc = &ccs->constraints[ccs->num_constraints++];
+	cc->fd.chunk_id = chunk_id;
+	cc->fd.dimension_slice_id = dimension_slice_id;
+
+	if (NULL == constraint_name)
+	{
+		if (is_dimension_constraint(cc))
+		{
+			snprintf(NameStr(cc->fd.constraint_name),
+					 NAMEDATALEN,
+					 "constraint_%d",
+					 cc->fd.dimension_slice_id);
+			namestrcpy(&cc->fd.hypertable_constraint_name, "");
+		}
+		else if (NULL != hypertable_constraint_name)
+		{
+			CatalogSecurityContext sec_ctx;
+			char		constrname[100];
```

Comment:
The static analyzer of the compiler will complain in `sprintf()` that the last arg (`hypertable_constraint_name`) might not fit since the serial is an `int64`. This is to silence that warning. The name will be truncated to NAMEDATALEN by `namestrcpy()`.

I would actually prefer we used the internal constraint naming machinery (`ChooseConstraintName()`, which is similar to `makeObjectName()`), but I figured I shouldn't change naming schemes, at least not for this PR.

Reply:
Btw, I am actually not sure why we need the `chunk_constraint` sequence at all. It seems sufficient to just have `chunk_id` and `constraint name` (i.e., it is a chunk-unique name). Then do conflict resolution by appending a unique label at the end (provided by both `makeObjectName() and `ChooseConstraintName()`). Now we don't have any conflict resolution at all.

Reply:
the sequence assures no conflict, but I agree, let's consider using another naming scheme in another PR.

Reply:
I don't see why changing 100->64 should give a warning, but if it does let's leave it.

Reply:
It does, as explained above. This is the warning: 
```
arning: ‘%s’ directive output may be truncated writing up to 63 bytes into a region of size between 31 and 60 [-Wformat-truncation=]
       "%d_" INT64_FORMAT "_%s"
```